### ISSUE TITLE
Streaming format should be a string not an integer

### ISSUE THREAD

**Issue by JonathanLennox** [moq-wg/catalog-format#44]
I think it would make more sense for the "streaming format" field to be a string rather than an integer.  (It should still be IANA-registered, though.)

**Comment by JonathanLennox** [comment]
(I think this was discussed in the chat at one of the interims, but it never got surfaced higher than that.)

**Comment by wilaw** [comment]
Seems reasonable. Should there be a max length? I find such fields get abused by people embedding 2,000 lines of javascript in the streaming format identifier.

**Comment by JonathanLennox** [comment]
Hm, that hadn't occurred to me, that's a good point...if it's IANA-registered you could require a registration policy that requires it to be something sensible, though this doesn't help for non-registered fields.

You'll probably want people to use identifiers like draft-my-proposed-new-catalog-format-04 for the draft versions that are under development (for the usual reason to prevent the draft versions from polluting the namespace for the final release version).  

Normally I'd invoke the computer science principle that "all arbitrary numbers should be powers of 2", but in this case it's possibly better to make it explicit that the value is indeed arbitrary, so I'll say:

Max 100 bytes.

I'm sure we could bikeshed forever about what characters should be allowed in the string, but I think that's something that needs to be defined more widely for this draft, since MoQTransport allows track namespaces, e.g., to be arbitrary binary strings.  I can open another issue for that.

As a strawman for this item specifically, I'd suggest ASCII-only, from the set `[A-Za-z0-9_.-]`.

**Comment by JonathanLennox** [comment]
Opened https://github.com/moq-wg/catalog-format/issues/45 for the track namespace grammar issue but I think the grammar of this field should be substantially more restricted.

**Comment by JonathanLennox** [comment]
Note for clarity: `[A-Za-z0-9_.-]` is standard regular expression syntax (Posix ERE or PCRE or whatever): square brackets are delimiters, `-` is a range operator except when at the beginning or the end.

So my suggestion is that tokens can contain alphanumerics, underscore, period, or dash, but nothing else.

**Comment by scottlamb** [comment]
How does the streaming format number (defined in section 3.2.1) relate to the packaging (defined in section 3.2.9)? It seems to be one-to-one? But the "mixed-format-catalog" example (section 3.4.8) seems to arbitrarily pick one format to declare at the catalog level, then declare the correct packaging at the track level.

What I'm getting at is: streaming format seems to be (a) redundant with the packaging, and (b) incorrectly declared at the catalog rather than track level. Should it be removed rather than turned into a string?

**Comment by wilaw** [comment]
> How does the streaming format number (defined in section 3.2.1) relate to the packaging (defined in section 3.2.9)? It seems to be one-to-one?

Not quite. A streaming format format can support multiple packaging formats. For example, WARP will support both CMAF and LOC packaging. The packaging just dictates how the audio and video is encapulsated. The encapsulation is one component of a streaming format. Signaling, naming, client and publisher behavior etc can be common between the different encapsulation types. 

![Screenshot 2024-05-02 at 9 38 54 AM](https://github.com/moq-wg/catalog-format/assets/2762250/5be6b51b-b66b-4d24-965d-4aab772ee022)

A WARP catalog would declare a streaming format number of "1" (or whatever we register it at IANA with) and then also must define the packaging type for each track, as it could be either CMAF or LOC. 

> But the "mixed-format-catalog" example (section 3.4.8) seems to arbitrarily pick one format to declare at the catalog level, then declare the correct packaging at the track level.

This example defines a streamingFormat of 1 (which indicates that WARP streaming format is being used) and then indicates that the video is packaged as CMAF but the audio track is packaged as Opus inside Low Overhead Container. This is likely to be a popular combination for non-DRM'd content, since the CMAF overhead inside low bitrate audio tracks can be unnecessarily large compared to LOC. 

```
{
  "version": 1,
  "streamingFormat": 1,
  "streamingFormatVersion": "0.2",
  "commonTrackFields": {
     "namespace": "output.example.com/event/12345",
     "renderGroup":1
  },
  "tracks": [
    {
      "name": "video0",
      "selectionParams":{"codec":"avc1.64001f","mimeType":"video/mp4","width":1280,"height":720,"framerate":30,"bitrate":4952892},
      "initTrack":"init_video_720",
      "packaging":"cmaf"
    },
    {
      "name": "audio",
      "selectionParams":{"codec":"opus","samplerate":48000,"channelConfig":"2","bitrate":32000},
      "packaging": "loc"
    }
   ]
}
```

**Comment by wilaw** [comment]
> Section 3.2.1 doesn't say that the streaming format defines "signaling, naming, client and publisher behavior"; should it?

It could and I guess there is no harm in adding that. At some point we need an overview document that describes how all the various specs interop to provide a streaming solution, basically a prose version of the [diagram](https://github.com/moq-wg/catalog-format/issues/44#issuecomment-2089813401). Not sure where that should reside?


> Section 3.4.8 is referred to as mixed "format" which I took to mean "streaming format". Perhaps it should just say "packaging" instead?

Indeed it should. Good catch.

[draft-law-moq-warpstreamingformat-latest](https://moq-wg.github.io/warp-streaming-format/draft-law-moq-warpstreamingformat.html) refers exclusively to CMAF. Should it indicate it can be used with other packaging formats?
Yes it should.  I am waiting for the LOC reference doc to be supplied to add in LOC packaging. But it would make sense to at least add a placeholder for LOC to make it clear that WARP is intended to support multiple packaging formats. I opened this [issue](https://github.com/moq-wg/warp-streaming-format/issues/24) for that purpose. 

>  it wasn't immediately obvious to me the packaging it declares was meant to be used independently of the streaming format it declares.

I agree. I asked Mo and Suhas if they would break that spec into two parts, one that references the packaging istelf and the other for the streaming format. See this [email ](https://mailarchive.ietf.org/arch/browse/moq/?q=Low%20Overhead%20Container%20(LOC)%20updated%20to%2003)on the moq list . If you agree, please chime in to that thread. 

> (btw, is there a later draft and/or github repo for this one?)

Not as of today. That's the latest.
